implementations/rust/ockam/ockam_transport_tcp/src/transport/lifecycle.rs

```diff
@@ -31,3 +43,138 @@ impl TcpTransport {
+                    ));
+                }
+
+                let options = if SocketAddr::from_str(address.address())
+                    .map(|socket_addr| socket_addr.ip().is_loopback())
+                    .is_ok()
+                {
+                    // TODO: Enable FlowControl for loopback addresses as well
+                    TcpConnectionOptions::insecure()
+                } else {
+                    let id = flow_controls.generate_id();
+                    TcpConnectionOptions::as_producer(flow_controls, &id)
+                };
+
+                number_of_tcp_hops += 1;
+                let addr = self.connect(address.address().to_string(), options).await?;
+                result = result.append(addr)
+            } else {
+                result = result.append(address.clone());
+            }
+        }
+
+        let resolved = result.into();
+        debug!("resolved route {} to {}", route, resolved);
+        Ok(resolved)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ockam_core::{route, TransportType, LOCAL};
+    use ockam_transport_core::TransportError;
+    use std::net::TcpListener;
+
+    #[ockam_macros::test]
+    async fn test_resolve_route(ctx: &mut Context) -> Result<()> {
+        let tcp = TcpTransport::create(ctx).await?;
+        let tcp_address = "127.0.0.1:5001";
+        let _listener = TcpListener::bind(tcp_address).map_err(TransportError::from)?;
+        let initial_workers = ctx.list_workers().await?;
+
+        let other_transport_type = TransportType::new(10);
+        let other_address = (other_transport_type, "other_address");
+        let route = tcp
+            .resolve_route(
+                &FlowControls::default(),
+                route![(TCP, tcp_address), other_address],
+            )
+            .await?;
+
+        // there are 2 additional workers
+        let mut additional_workers = ctx.list_workers().await?;
+        additional_workers.retain(|w| !initial_workers.contains(w));
+        assert_eq!(additional_workers.len(), 2);
+
+        // the TCP address is replaced with the TCP sender worker address
+        // the other address is left unchanged
+        assert_eq!(
+            route.iter().map(|a| a.transport_type()).collect::<Vec<_>>(),
+            vec![LOCAL, other_transport_type]
+        );
+
+        let first_address = route.next()?;
+        assert!(additional_workers.contains(first_address));
+
+        // trying to resolve the address a second time should still work
+        let _route = tcp
+            .resolve_route(
+                &FlowControls::default(),
+                route![(TCP, tcp_address), other_address],
+            )
+            .await?;
+
+        ctx.stop().await
+    }
+
+    #[ockam_macros::test]
+    async fn test_resolve_route_only_single_hop_is_allowed(ctx: &mut Context) -> Result<()> {
+        let tcp = TcpTransport::create(ctx).await?;
+        let tcp_address = "127.0.0.1:5002";
```

Comment:
let tcp_address = "127.0.0.1:0";

Reply:
Thanks, I see what you mean now, and it works fine